com.unity.netcode.gameobjects/Runtime/Utility/RefArray.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Unity.Multiplayer.Netcode
+{
+    public ref struct RefArray<T> where T : unmanaged
```

Comment:
We do actually support readonly NativeArrays :D you can just: 

```
var x = new NativeArray<byte>(1, Allocator.Temp);
var ro = x.AsReadOnly(); // this returns a readonly view into the array
```

Reply:
Also we have a Span<T> type for native collections its called NativeSlice<T> and it can do a lot the same things Span can.

Reply:
Hey all! Could someone help me understand what is the actual difference between these containers/wrappers and the Native ones? I am trying better understand the thinking behind why we felt like we needed them? Personally - and as Andrew mentioned - it feels like these use cases would be already covered by Native Collections.

Reply:
Being read-only isn't the point... The point is passing something created with stackalloc without copying it. NativeArray would create a new temp allocation and copy, which would defeat the purpose of the original stackalloc. It also doesn't address that NativeArray is slower than RefArray due to all its safeguards (and still slower even in release builds), and that NativeArray doesn't implement IReadOnlyList, which I will need.

I can fix the second with a wrapper class, but that'll likely make the first worse. But maybe I'm over-optimizing here? RefArray is a simple enough struct that I think it's worthwhile. 

Also NativeSlice doesn't replace Span because it requires a NativeArray to construct it. The reason I mentioned Span is that in mainline C# the compiler allows this: 

```c#
Span<int> foo = stackalloc int[5]();
```

Unity has no type that can be directly assigned to from a stackalloc to be able to pass the stackalloc'd memory to another function. That's the use case for RefArray - neither NativeArray nor NativeSlice enable that usage.

Reply:
Oh, I see why you mentioned being read-only. I don't need IReadOnlyList to make it read only. I need IReadOnlyList because it's the interface that provides array.Count and array[idx] for generic functions. NativeArray has both of those, but doesn't list IReadOnlyList among its implemented interfaces, so I can't use it for this use. I don't actually care if it's read only, I care if it's able to be used in a generic function, and NativeArray is not. It does let me write a generic function that uses the `foreach` syntax, but that causes boxing allocations converting the returned enumerator to IEnumerator when used in a generic context.

Reply:
Tl;dr for Cosmin... The difference is, RefArray

- Can wrap an existing pointer (such as a return from stackalloc) __without any copies__
- Has less overhead since it's not intended for thread safety in any way and its functions are all aggressively inlined
- Can be used generically without incurring boxing allocations

None of those are true for NativeArray. NativeArray has a lot of good uses, but doesn't fit the specific need I'm trying to address with RefArray.

Reply:
To put it another way, RefArray exists to give the IReadOnlyList interface to a raw pointer and pass that pointer to a generic function. Very specific use case. Not intended to replace NativeArray for any other usage.

Reply:
> Can wrap an existing pointer (such as a return from stackalloc) without any copies

You can do: 

```
int* foo = stackalloc int[5]();
var xx = NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<int>(foo, 5, Allocator.Temp);
```
This doesn't do a copy at all.

Reply:
Ok. That would work for this use case - in my defense, though, that function's hard to find. :P

I still contend that RefArray will be faster than NativeArray, but maybe not enough for us to care in this case. 🤷 And I'll still have to write a wrapper for NativeArray to add `IReadOnlyList` support, so the difference would end up being

```c#
int* foo = stackalloc int[5]();
var xx = new RefArray(foo, 5);
```

vs

```c#
int* foo = stackalloc int[5]();
var xx = new NativeArrayReadOnlyListWrapper(NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<int>(foo, 5, Allocator.None));
```

The former is still a lot more readable to me, in addition to the performance gains.

But I'll remove it if you still want me to after having heard my reasoning.

Reply:
why do we care about `IReadOnlyList` interface this much again? 👀

Reply:
`public void SendMessage<T, U>(in T message, in U clientIds) where T: IMessage where U: IReadOnlyList<ulong>`

Allows passing whatever you have readily available. If you have `ulong[]` or `List<ulong>` already, you can pass it as is. If you need to create something, you can `stackalloc` and pass a wrapper with no allocations. Taking `ulong[]` forces allocating any time the list we need doesn't already happen to exist. 

`IReadOnlyList` is important because it supplies `array.Count` and `array[index]`. NativeArray implements IEnumerable. If that function took IEnumerable instead of IReadOnlyList, you could still pass all of those types, but every time you call `foreach` on it, you would get a boxing allocation because it has to convert the returned value to IEnumerable. Any polymorphism that isn't on a generic type parameter with a where clause will always result in boxing. Can't be zero-alloc if we box every time we send.

Reply:
We can get around that boxing with IReadOnlyList by using a c-style for loop and using index access instead of using foreach.

Reply:
After talking with Jesse, I think we've come to a good compromise on this. I'm deleting this file for now. A more focused implementation to enable the same concept will show up in the IMessage PR in a way that I think will be more palatable.